clippy_lints/src/ptr_offset_by_literal.rs

```diff
@@ -0,0 +1,156 @@
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        // Check if the expressions is a ptr.offset or ptr.wrapping_offset method call
+        let Some((receiver_expr, arg_expr, method)) = expr_as_ptr_offset_call(cx, expr) else {
+            return;
+        };
+
+        // Check if the argument to the method call is a (negated) literal.
+        let Some(literal) = expr_as_literal(arg_expr) else {
+            return;
+        };
+
+        let msg = format!("use of `{method}` with a literal");
+        if let Some(sugg) = build_suggestion(cx, method, receiver_expr, literal) {
+            span_lint_and_sugg(
+                cx,
+                PTR_OFFSET_BY_LITERAL,
+                expr.span,
+                msg,
+                "try",
+                sugg,
+                Applicability::MachineApplicable,
+            );
+        } else {
+            span_lint(cx, PTR_OFFSET_BY_LITERAL, expr.span, msg);
+        }
+    }
+}
+
+fn get_literal_bits<'tcx>(expr: &'tcx Expr<'tcx>) -> Option<u128> {
+    let ExprKind::Lit(lit) = expr.kind else {
+        return None;
+    };
+
+    let LitKind::Int(packed_u128, _) = lit.node else {
+        return None;
+    };
+
+    Some(packed_u128.get())
+}
+
+// If the given expression is a (negated) literal, return its value.
+fn expr_as_literal<'tcx>(expr: &'tcx Expr<'tcx>) -> Option<i128> {
+    if let Some(literal_bits) = get_literal_bits(expr) {
+        // The value must fit in a isize, so we can't have overflow here.
+        return Some(literal_bits as i128);
+    }
+
+    if let ExprKind::Unary(UnOp::Neg, inner) = expr.kind {
+        if let Some(literal_bits) = get_literal_bits(inner) {
+            return Some(-1 * literal_bits as i128);
+        }
+    }
+
+    None
+}
+
+// If the given expression is a ptr::offset  or ptr::wrapping_offset method call, return the
+// receiver, the arg of the method call, and the method.
+fn expr_as_ptr_offset_call<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'_>,
+) -> Option<(&'tcx Expr<'tcx>, &'tcx Expr<'tcx>, Method)> {
+    if let ExprKind::MethodCall(path_segment, arg_0, [arg_1], _) = &expr.kind
+        && cx.typeck_results().expr_ty(arg_0).is_raw_ptr()
+    {
+        if path_segment.ident.name == sym::offset {
+            return Some((arg_0, arg_1, Method::Offset));
+        }
+        if path_segment.ident.name == sym::wrapping_offset {
+            return Some((arg_0, arg_1, Method::WrappingOffset));
+        }
+    }
+    None
+}
+
+fn build_suggestion(cx: &LateContext<'_>, method: Method, receiver_expr: &Expr<'_>, literal: i128) -> Option<String> {
+    let receiver = receiver_expr.span.get_source_text(cx)?;
+
+    let new_method = match Ord::cmp(&literal, &0) {
+        Ordering::Greater => match method {
+            Method::Offset => "add",
+            Method::WrappingOffset => "wrapping_add",
+        },
+        Ordering::Equal => return Some(format!("{receiver}")),
```

Comment:
This case looks a bit surprising, maybe add a short comment to it? Something like:
```suggestion
        // `ptr.offset(0)` is equivalent to `ptr`, so no adjustment is needed
        Ordering::Equal => return Some(format!("{receiver}")),
```
(why on earth do the suggestions look like this)